pyteal/ast/subroutine.py

```diff
@@ -19,7 +18,10 @@ class SubroutineDefinition:
     nextSubroutineId = 0
 
     def __init__(
-        self, implementation: Callable[..., Expr], returnType: TealType
+        self,
+        implementation: Callable[..., Expr],
+        returnType: TealType,
+        methodSignature: str = None,
```

Comment:
I'm less sure about the changes to this file, since not every subroutine will represent a method.

What if you changed `methodSignature` to just `name`? I.e. if you provide the `name` arg, then it would override whatever the function's `__name__` is.

@barnjamin thoughts on this?

Reply:
I had thought we'd be able to infer the method signature from the method definition, including name and types it accepts. That would mean defining the ABI types with accessors.

Reply:
I started something here https://github.com/barnjamin/pyteal-utils/blob/main/pytealutils/abi/types.py but maybe this kind of thing belongs in PyTEAL proper.

We should _definitely_ have some representation of them so decoding from the over-the-wire format is painless

Reply:
I ended up just adding the ABI types to utils for now using the py-sdk to test encoding/decoding:  https://github.com/algorand/pyteal-utils/blob/a5f35c9e217b2c6e259b8a725bb988ec631e6fc8/pytealutils/abi/test_fixed_point.py

Reply:
Just changed `methodSignature` name to `nameStr`, and override function's `__name__` on initialization, but I still feel this needs to be polished as an implementation, and should be modified after ABI in PyTeal is introduced.

I would like to keep this thread open for more discussion on the design of subroutine (in combination with ABI methods).